tempodb/blocklist/poller.go

```diff
@@ -148,44 +150,69 @@ func (p *Poller) Do(previous *List) (PerTenant, PerTenantCompacted, error) {
 		return nil, nil, err
 	}
 
-	blocklist := PerTenant{}
-	compactedBlocklist := PerTenantCompacted{}
+	var (
+		wg  = boundedwaitgroup.New(p.cfg.TenantPollConcurrency)
+		mtx = sync.Mutex{}
+
+		blocklist          = PerTenant{}
+		compactedBlocklist = PerTenantCompacted{}
 
-	consecutiveErrors := 0
+		finalErr          atomic.Error
+		consecutiveErrors int
+	)
 
 	for _, tenantID := range tenants {
-		newBlockList, newCompactedBlockList, err := p.pollTenantAndCreateIndex(ctx, tenantID, previous)
-		if err != nil {
-			level.Error(p.logger).Log("msg", "failed to poll or create index for tenant", "tenant", tenantID, "err", err)
-			consecutiveErrors++
-			if consecutiveErrors > p.cfg.TolerateConsecutiveErrors {
-				level.Error(p.logger).Log("msg", "exiting polling loop early because too many errors", "errCount", consecutiveErrors)
-				return nil, nil, err
+		// The finalErr is only stored when the number of consecutive errors exceeds the tolerance.  If present, return it.
+		if err := finalErr.Load(); err != nil {
+			level.Error(p.logger).Log("msg", "exiting polling loop early because too many errors", "errCount", consecutiveErrors)
+			return nil, nil, err
+		}
+
+		wg.Add(1)
+		go func(tenantID string) {
+			defer wg.Done()
+
+			newBlockList, newCompactedBlockList, err := p.pollTenantAndCreateIndex(ctx, tenantID, previous)
```

Comment:
Maybe we could do something like:

```go
newBlockList, newCompactedBlockList, err := p.pollTenantAndCreateIndex(ctx, tenantID, previous)
resultCh <- result{
		tenantID:           tenantID,
		blockList:          newBlockList,
		compactedBlockList: newCompactedBlockList,
		err:                err,
}
```

And handle the errors outside the coroutine to avoid the mutex?

Reply:
i personally like the simplicity of this approach. it's quite easy to read. i wouldn't be opposed to a struct through a channel if we really felt we needed it.